std/algorithm/iteration.d

```diff
@@ -888,66 +889,70 @@ template each(alias pred = "a")
         (!isForwardRange!R || isDynamicArray!R) &&
         (isForeachUnaryIterable!R || isForeachBinaryIterable!R);
 
+public:
     void each(Range)(Range r)
-    if (isRangeIterable!Range && !isForeachIterable!Range)
+    if (!isForeachIterable!Range && (
+        isRangeIterable!Range ||
+        __traits(compiles, typeof(r.front).length)))
     {
-        debug(each) pragma(msg, "Using while for ", Range.stringof);
-        static if (isRangeUnaryIterable!Range)
+        static if (isRangeIterable!Range)
         {
-            while (!r.empty)
+            debug(each) pragma(msg, "Using while for ", Range.stringof);
+            static if (isRangeUnaryIterable!Range)
             {
-                cast(void) unaryFun!pred(r.front);
-                r.popFront();
+                while (!r.empty)
+                {
+                    cast(void) unaryFun!pred(r.front);
+                    r.popFront();
+                }
             }
-        }
-        else // if (isRangeBinaryIterable!Range)
-        {
-            size_t i = 0;
-            while (!r.empty)
+            else // if (isRangeBinaryIterable!Range)
             {
-                cast(void) binaryFun!BinaryArgs(i, r.front);
-                r.popFront();
-                i++;
+                size_t i = 0;
+                while (!r.empty)
+                {
+                    cast(void) binaryFun!BinaryArgs(i, r.front);
+                    r.popFront();
+                    i++;
+                }
             }
         }
+        else
+        {
+            // range interface with >2 parameters.
+            for (auto range = r; !range.empty; range.popFront())
+                pred(range.front.expand);
+        }
     }
 
     void each(Iterable)(auto ref Iterable r)
-        if (isForeachIterable!Iterable)
+    if (isForeachIterable!Iterable ||
+        __traits(compiles, Parameters!(Parameters!(r.opApply))))
```

Comment:
The code for the second group checks out as well (assuming you meant they correspond with the 2nd and 3rd overloads in the original).